Skale-contracts integration

Masterix0 opened the pull request:
### Overview

This pull request introduces a major overhaul of the environment configuration and validation in node-cli. The key changes include:

- **Simplified Environment Variable Handling:**  
  - Replaces the previous ABI JSON download mechanism (using `MANAGER_CONTRACTS_ABI_URL` and `IMA_CONTRACTS_ABI_URL`) with a new approach that uses the environment variables `MANAGER_CONTRACTS` and `IMA_CONTRACTS`.  
  - The new approach forwards these values (which can be either direct contract addresses or recognized aliases) to other SKALE network components that use the skale-contracts library for dynamic ABI access.

- **Refactored Validation Logic:**  
  - The `env.py` file has been refactored to introduce helper functions to modularize the process of loading and validating environment variables.  
  - Validation now explicitly checks the new variables (checking if alias is present in skale repo or if contract addresses have code in provided endpoint), ensuring that the environment is correctly configured before node-cli proceeds.
  - Deprecated validation code related to ABI JSON files has been removed.

- **Enhanced Testing:**  
  - A new test suite (`tests/configs/configs_env_validate_test.py`) has been created to extensively test the updated environment functions.  
  - External HTTP requests are patched using `requests_mock` to simulate network responses, ensuring that the tests remain isolated.

- **Workflow and Documentation Improvements:**  
  - The README has been updated to reflect the new environment variable names and usage instructions.  
  - The GitHub Actions workflow now includes a new step to run `ruff` for linting and code formatting, improving our code quality checks.
  - Adjustments were made to the build process (specifically the generation of `info.py`) to avoid development-time import errors.

### Detailed Changes

- **Environment Configuration Updates:**  
  - **New Variables:**  
    - Replace `MANAGER_CONTRACTS_ABI_URL` and `IMA_CONTRACTS_ABI_URL` with `MANAGER_CONTRACTS` and `IMA_CONTRACTS`.  
    - These variables now contain either a direct contract address or an alias.
  - **Refactoring:**  
    - Split the environment loading into three parts: checking and loading the file (`load_env_file()`), building a base parameters dictionary (`build_params()`), and populating it with environment values (`populate_params()`).
  - **Validation:**  
    - `validate_params()` now calls `validate_env_type()` for environment type variable and `validate_env_alias_or_address()` for both the IMA and Manager contracts using the new variables.
    - Online validations (e.g., fetching the chain ID and network metadata) are performed only after basic checks have passed.

- **Testing Enhancements:**  
  - A dedicated test suite for the new environment configuration is added, covering:
    - Missing or unreadable env files.
    - Correct population of parameters from the environment.
    - Successful and failing validations for environment types, contract addresses, and aliases.
  - External HTTP calls are mocked using `requests_mock` to isolate tests from real network dependencies.

- **Workflow & Linting:**  
  - Integrated `ruff` into our CI/CD pipelines via GitHub Actions.
  - Updated the README with instructions regarding the new environment variables and the use of `ruff`.
  
- **Build Process Improvement:**  
  - The generation of `info.py` is now handled in a separate script (referenced in the README and CI), reducing development-time import errors.

### Impact

- **Breaking Changes:**  
  - Scripts or integrations relying on the old ABI URL environment variables will need to be updated to use the new alias/contract address variables.
  - The deprecated `validate` command group has been removed.

### Testing

- All unit and integration tests for environment configuration now pass.
- CI/CD has been updated to run `ruff` for code style and formatting checks and to only run test pipeline on `push`.

Masterix0 committed:
Corrected optional monitoring containers on README

Masterix0 committed:
Updated .env variables in README

Masterix0 committed:
Updated README to be more clear and up to date.

Masterix0 committed:
Add files via upload

Added sample info.py file to fix pylance error since file doesn't existe.
File will be overwritten at build time.

Masterix0 committed:
Updated build script to make sure info.py file behaves like before.

Masterix0 committed:
Fixed erroneous information in README

Masterix0 committed:
Add skale-contracts dependency to setup.py

Masterix0 committed:
Refactor node-cli to use skale-contracts for contract alias/address validation

• Introduced dynamic fetching of contract ABIs from skale-contracts by allowing either an alias (e.g., “test-manager”) or a direct contract address (e.g., “0xabc123…”) in the environment.
• Renamed and replaced the old environment variables that directly pointed to local ABI files.
• Created new validation logic (validate_env_alias_or_address) for both Manager and IMA contracts in the CLI.
• Removed the obsolete tests that ensured local ABI files existed/were valid JSON and replaced them with new tests that check for valid aliases or addresses.
• Updated existing tests and mocks to patch or skip the new external validation calls where necessary.
• Added ruff configuration (ruff.toml) and reformatted code throughout the repository.

Masterix0 committed:
Fix formatting issues in node_test.py causing flake8 to fail

Masterix0 committed:
Refactored info.py generation into separate script.
Added it to README dev preparation so there are no problems with missing info.py.
Added info.py to .gitignore.

Masterix0 committed:
Add contract address validation utility function in helper.py and update contract address validation logic to use this utility function.

Masterix0 committed:
Renamed 'MANAGER_CONTRACTS_ALIAS_OR_ADDRESS' and 'IMA_CONTRACTS_ALIAS_OR_ADDRESS' to 'MANAGER_CONTRACTS' and 'IMA_CONTRACTS', respectively.

Masterix0 committed:
Add ruff check to pre-commit hook indication in README

dmytrotkk commented on README.md:
```diff
@@ -111,22 +110,26 @@ Arguments:
+- `DOCKER_LVMPY_STREAM` - stream of `docker-lvmpy` to use
+- `CONTAINER_CONFIGS_STREAM` - stream of `skale-node` to use
+- `ENDPOINT` - RPC endpoint of the node in the network where SKALE Manager is deployed
+- `MANAGER_CONTRACTS` - SKALE Manager main contract alias or address
```
makes sense to specify which one is the main contract in skale-manager and which one is the main in IMA

dmytrotkk commented on README.md:
```diff
@@ -111,22 +110,26 @@ Arguments:
+- `TG_CHAT_ID` - Telegram chat ID
+- `MONITORING_CONTAINERS` - will enable monitoring containers (`cadvisor`, `node-exporter`).
+
+> Filebeat is always enabled and requires `FILEBEAT_URL`, it is **not optional**.
```
The correct variable name is `FILEBEAT_HOST`, also this note can be removed, since `FILEBEAT_HOST` is already listed as a required variable.

dmytrotkk commented on .github/workflows/test.yml:
```diff
@@ -30,10 +30,17 @@ jobs:
+    - name: Generate info
+      run: ./scripts/generate_info.sh 1.0.0 my-branch normal
+
     - name: Lint with flake8
```
Since we moved to `ruff`, let's remove flake from pipelines, dependencies and readme

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -1,38 +1,56 @@
+"""Environment configuration and validation module for SKALE node.
```
replace with the correct license header

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -45,36 +63,170 @@
+    """Check and load environment variables from the given file."""
+    if not os.path.exists(env_filepath):
+        error_exit(f'Environment file not found: {env_filepath}', CLIExitCodes.FAILURE)
+    if not os.access(env_filepath, os.R_OK):
```
validations could be simplified or removed - one `load_dotenv` covers all cases.
also, `dotenv_values` can be used to check .env file without actually loading variables.

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -45,36 +63,170 @@
+        validate_contract_alias(alias_or_address, contract_type, endpoint)
+
+
+def validate_contract_address(contract_address: str, endpoint: str) -> None:
```
move everything related to the alias/address validations to the separate module

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -45,36 +63,170 @@
+    if missing:
+        error_exit(f'Missing required parameters: {missing}', CLIExitCodes.FAILURE)
+    validate_env_type(params['ENV_TYPE'])
+    # Get the endpoint explicitly from the params.
```
remove the comment

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -45,36 +63,170 @@
+
+def get_network_metadata() -> Dict:
+    """Fetch network metadata from GitHub."""
+    metadata_url = (
```
move to constants

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -45,36 +63,170 @@
+    alias_or_address: str, contract_type: ContractType, endpoint: str
+) -> None:
+    """Validate contract alias or address."""
+    if not alias_or_address:
```
this validation can be removed, since empty `IMA_CONTRACTS` or `MANAGER_CONTRACTS` will be handled in `absent_params(params)`

Masterix0 committed:
Merge remote-tracking branch 'origin/develop' into skale-contract-integration

Masterix0 committed:
Updated helper-scripts version

dmytrotkk commented on node_cli/cli/__init__.py:
```diff
@@ -1,4 +1,4 @@
-__version__ = '2.6.0'
+__version__ = '2.6.2'
```
bump version to 3.0.0

dmytrotkk committed:
Merge branch 'v4.1.x' of github.com:skalenetwork/node-cli into skale-contract-integration

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -45,36 +63,170 @@
+            error_exit(
+                f'Network with chain ID {chain_id} not found in metadata', CLIExitCodes.FAILURE
+            )
+        deployment_url = (
```
mb let's move the code to`get_deployment_url` function (optional)

dmytrotkk commented on node_cli/core/host.py:
```diff
@@ -65,52 +74,91 @@ def fix_url(url):
-    return os.path.isfile(RESOURCE_ALLOCATION_FILEPATH)
+
+def is_node_inited() -> bool:
+    """Check if the SKALE node has been initialized.
```
comment can be removed since the code is self-explanatory, also try/except can be dropped

dmytrotkk commented on node_cli/core/host.py:
```diff
@@ -65,52 +74,91 @@ def fix_url(url):
-    copyfile(env_filepath, SKALE_DIR_ENV_FILEPATH)
+def save_env_params(env_filepath: str) -> None:
+    """Copy environment parameters file to SKALE directory."""
+    if not os.path.isfile(env_filepath):
```
https://www.reddit.com/r/programming/comments/7xyt5d/overly_defensive_programming/

> If your exception handling makes your code difficult to read, or to have so much boilerplate that people won't read large sections of it, you have destroyed the readability of your code.
Why is readability more important than an obvious bug like swallowing an exception? Bugs can be patched, but code that is hard to read has to be rewritten, which is a much bigger hassle.

dmytrotkk commented on node_cli/core/node.py:
```diff
@@ -223,35 +220,45 @@ def cleanup_sync() -> None:
-    sync_schains=None,
-    pull_config_for_schain=None,
-    sync_node=False,
+    env_filepath: Optional[str],
```
`env_filepath` is not optional

dmytrotkk commented on node_cli/core/node.py:
```diff
@@ -223,35 +220,45 @@ def cleanup_sync() -> None:
-        env_params = extract_env_params(INIT_ENV_FILEPATH, sync_node=sync_node)
+        env_params = get_env_config(INIT_ENV_FILEPATH, sync_node=sync_node)
 
+    # Set mount directory based on node type
```
remove redundant comments (same for the rest of the file)

dmytrotkk commented on node_cli/operations/skale_node.py:
```diff
@@ -43,17 +40,35 @@ def update_images(env: dict, sync_node: bool = False) -> None:
+                )
+            logger.info(f'Syncing config files from {src}')
+            rsync_dirs(src, dest)
+        elif stream:
```
let's keep a simple if/else here since you've already added a check for src/stream at the beginning of the function

dmytrotkk commented on node_cli/main.py:
```diff
@@ -123,5 +123,5 @@ def handle_exception(exc_type, exc_value, exc_traceback):
         traceback.print_exc()
         logger.debug('Execution time: %d seconds', time.time() - start_time)
-        error_exit(err)
+        error_exit(err, CLIExitCodes.FAILURE)
```
`CLIExitCodes.FAILURE` is already the default option for the error code, no need to explicitly provide it

dmytrotkk commented on node_cli/utils/git_utils.py:
```diff
@@ -21,29 +21,65 @@
 
 
 def check_is_branch(repo: Repo, ref_name: str) -> bool:
+    """Check if the given reference name is a valid git branch."""
```
suggest to simplify this function - drop checks, optimize the code.
it can be as simple as that:

```python
def check_is_branch(repo: Repo, ref_name: str) -> bool:
    return ref_name in (branch.name for branch in repo.heads)
```

dmytrotkk commented on node_cli/utils/git_utils.py:
```diff
@@ -21,29 +21,65 @@
-    Repo.clone_from(repo_url, repo_path)
-    fetch_pull_repo(repo_path, ref_name)
+    """Clone a git repository and checkout specified reference."""
+    if not all([repo_url, repo_path, ref_name]):
```
is there a place in the code where one of those arguments is not provided?
if not, remove the check

dmytrotkk commented on node_cli/utils/git_utils.py:
```diff
@@ -21,29 +21,65 @@
+    """Clone a git repository and checkout specified reference."""
+    if not all([repo_url, repo_path, ref_name]):
+        error_exit('Empty repository URL, path or reference', CLIExitCodes.FAILURE)
+    if not all(isinstance(x, str) for x in [repo_url, repo_path, ref_name]):
```
same for type check - is it realistically possible in the code that values will be provided, but will have wrong type

dmytrotkk commented on node_cli/utils/git_utils.py:
```diff
@@ -52,11 +88,32 @@ def sync_repo(repo_url: str, repo_path: str, ref_name: str) -> None:
+        repo = Repo(repo_path)
+        repo_name = os.path.basename(repo.working_dir)
+
+        # Fetch latest changes
```
remove unnecessary comments, also check if it makes sense to drop some of the checks/error handlers

dmytrotkk commented on node_cli/utils/helper.py:
```diff
@@ -354,20 +338,52 @@ def rm_dir(folder: str) -> None:
 
-def safe_mkdir(path: str, print_res: bool = False):
+def safe_mkdir(path: str, print_res: bool = False) -> None:
+    """Create a directory if it doesn't exist."""
```
add docstrings only for complex functions with unclear behavior

dmytrotkk commented on node_cli/utils/helper.py:
```diff
@@ -354,20 +338,52 @@ def rm_dir(folder: str) -> None:
-def safe_mkdir(path: str, print_res: bool = False):
+def safe_mkdir(path: str, print_res: bool = False) -> None:
+    """Create a directory if it doesn't exist."""
+    if not isinstance(path, str):
```
is it possible in the code that this function will get a non-string input?
if no, remove the check

dmytrotkk commented on node_cli/utils/helper.py:
```diff
@@ -354,20 +338,52 @@ def rm_dir(folder: str) -> None:
-    logger.info(f'Syncing {dest} with {src}')
-    run_cmd(['rsync', '-r', f'{src}/', dest])
-    run_cmd(['rsync', '-r', f'{src}/.git', dest])
+    """Synchronize two directories using rsync."""
```
docstring and function name are basically identical

dmytrotkk commented on node_cli/utils/print_formatters.py:
```diff
@@ -284,20 +242,41 @@ def print_node_info(node, node_status):
+            # Join list items with newlines for multiple errors
+            error_msg = '\n'.join(str(err) for err in error_payload)
+        elif isinstance(error_payload, dict):
+            # Format dict as JSON string
```
remove unnecessary comments

dmytrotkk commented on tests/configs/configs_env_validate_test.py:
```diff
@@ -0,0 +1,323 @@
+
+
+# =============================================================================
+# Helper fake response for patching requests.get in network helpers
```
remove comments from test functions/classes

dmytrotkk commented on tests/configs/configs_env_validate_test.py:
```diff
@@ -0,0 +1,323 @@
+# =============================================================================
+# Tests for absent_params
+# =============================================================================
+class TestAbsentParams:
```
all tests in skale.py, node-cli, skale-admin, and other repos use a function-based approach for pytest, not classes.
also, the pytest was designed with function-based tests as the primary pattern.

update new tests to use isolated functions instead of classes
if functions in a class do not use any shared resources and are just used to separate logical pieces, it make sense to move them to the separate files instead

Masterix0 committed:
Address review comments for PR #839

Masterix0 committed:
Removed unused imports

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -1,38 +1,68 @@
+from typing import Dict, List, Optional
 from dotenv import load_dotenv
-from node_cli.configs import SKALE_DIR, CONTAINER_CONFIG_PATH
+import requests
```
wrong order of imports
should be:

```python
import os
from enum import Enum

from typing import Dict, List, Optional

import requests
from dotenv import load_dotenv
```

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -155,16 +162,20 @@ def validate_contract_address(contract_address: str, endpoint: str) -> None:
+
+
+def get_deployment_url(alias: str, contract_type: ContractType, network_path: str) -> str:
+    """Construct the deployment URL for the given contract alias and type."""
```
obvious comment for a simple single line function

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -203,27 +207,23 @@ def get_chain_id(endpoint: str) -> int:
     except requests.RequestException as e:
-        error_exit(f'Failed to get chain ID: {str(e)}', CLIExitCodes.FAILURE)
+        error_exit(f'Failed to get chain ID: {str(e)}')
         # Will never reach this line, but needed for type checking.
```
remove all comments and return statements for "# Will never reach this line, but needed for type checking."

if you want type checks to work correctly, update return type of `error_exit` function to `NoReturn` from `typing`

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -112,9 +126,8 @@ def validate_params(params: Dict[str, str]) -> None:
     missing = absent_params(params)
     if missing:
-        error_exit(f'Missing required parameters: {missing}', CLIExitCodes.FAILURE)
+        error_exit(f'Missing required parameters: {missing}')
```
function name: `validate params`
docstring: `this function validates params`

which additional context does it provide except making the code less readable?

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -123,19 +136,13 @@ def validate_params(params: Dict[str, str]) -> None:
 def validate_env_type(env_type: str) -> None:
```
function name: `validate env type`
docstring: `Validate the environment type`

> Use comments to explain complex or non-obvious code, but avoid unnecessary comments that simply restate what the code does.

suggest to remove all docstrings from this module

dmytrotkk commented on node_cli/configs/env.py:
```diff
@@ -179,20 +190,13 @@ def validate_contract_alias(alias: str, contract_type: ContractType, endpoint: s
-            f'refs/heads/deployments/{network_path}/{contract_type.value}/{alias}.json'
-        )
+            error_exit(f'Network with chain ID {chain_id} not found in metadata')
+        assert isinstance(network_path, str)
```
do not use asserts outside of tests

https://www.reddit.com/r/learnpython/comments/zd217y/comment/iyzfq5l/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

Masterix0 committed:
Removes unnecessary 'self' parameter from test functions

dmytrotkk commented on tests/configs/configs_env_validate_test.py:
```diff
@@ -34,290 +31,249 @@ def json(self):
+    os.chmod(env_file, 0o644)  # reset permissions
+
+
+def test_build_params_sync(self):
```
`test_build_params_sync` and `test_build_params_non_sync` can be combined in one test

dmytrotkk commented on tests/configs/configs_env_validate_test.py:
```diff
@@ -34,290 +31,249 @@ def json(self):
+
+
+def test_validate_contract_address_success(self, requests_mock):
+    endpoint = 'http://localhost:8545'
```
suggest to move to constants (at top of the file)

Masterix0 committed:
Added exception handling during environment file loading.

dmytrotkk commented on node_cli/core/host.py:
```diff
@@ -65,52 +73,71 @@ def fix_url(url):
+            return secret_key
+    except (IOError, OSError) as e:
+        error_exit(f'Failed to read Flask secret key: {e}')
+        # Will never reach here, but needed for type checking.
```
remove this part (see `NoReturn` comment above)

dmytrotkk commented on node_cli/utils/helper.py:
```diff
@@ -418,3 +412,8 @@ def get_ssh_port(ssh_service_name='ssh'):
+
+# TODO: Add a more robust check for contract address and see if doesn't conflict with alias.
+def is_contract_address(value: str) -> bool:
+    return len(value) == 42 and value.startswith('0x')
```
can replace with `bool(re.fullmatch(r"0x[a-fA-F0-9]{40}", value))` and remove comment

dmytrotkk replied:
```
>>> value = '0x83b7b3A59857188474FF0269cCC3859b4647UMPQ'
>>> len(value) == 42 and value.startswith('0x')
True
>>> bool(re.fullmatch(r"0x[a-fA-F0-9]{40}", value))
False
```

dmytrotkk commented on tests/configs/configs_env_validate_test.py:
```diff
@@ -0,0 +1,279 @@
+
+def test_load_env_file_nonexistent():
+    with pytest.raises(SystemExit) as excinfo:
+        load_env_file('nonexistent.env')
```
suggest keeping `load_env_file` like this:

```python
def load_env_file(env_filepath: str) -> None:
    if not load_dotenv(dotenv_path=env_filepath):
        error_exit(f'Failed to load environment from {env_filepath}')
```

and if `load_dotenv` will crash, allow it to do it - exception will be handled on the top level (`node_cli/main.py`)

badrogger commented on tests/conftest.py:
```diff
@@ -307,6 +294,67 @@ def tmp_sync_datadir():
+        for key, value in valid_env_params.items():
+            f.write(f'{key}={value}\n')
+        file_name = f.name
+    yield file_name
```
Better to add try/finally here

badrogger commented on node_cli/core/host.py:
```diff
@@ -65,52 +73,71 @@ def fix_url(url):
-        update_resource_allocation(env_type)
+        if allocation:
+            update_resource_allocation(env_type)
+    except Exception as e:
```
It is cleaner to handle base Exception class in one place. It is already handled in main.py. Suggest to raise custom exception type instead and run error_exit in main.py with specific error index.

Masterix0 committed:
Addresses comments from PR#839

- Refactored environment validation into two files;
- Updated environment validation function names for improved clarity;
- Fixed minor docstring inconsistencies and formatting issues across multiple modules.

Masterix0 committed:
Fixes typo in resource allocation file debug message

Masterix0 committed:
Adjusted tests to expect PermissionError for unreadable files & Updated test pipeline to only run tests on push

dmytrotkk commented on node_cli/operations/skale_node.py:
```diff
@@ -43,17 +40,31 @@ def update_images(env: dict, sync_node: bool = False) -> None:
+            logger.info(f'Syncing config files from {src}')
+            rsync_dirs(src, dest)
+        else:
+            assert stream
```
what this line does?

also: do not use `assert` outside of tests

Masterix0 replied:
It's needed by type checking. Replaced with `if not: error_exit()`

dmytrotkk commented on node_cli/configs/alias_address_validation.py:
```diff
@@ -0,0 +1,118 @@
+#
+#   This file is part of node-cli
+#
+#   Copyright (C) 2019-Present SKALE Labs
```
should be `2025-Present` for the new file

dmytrotkk commented on node_cli/utils/git_utils.py:
```diff
@@ -21,29 +21,29 @@
-    logger.info(f'Cloning {repo_url} → {repo_path}')
-    Repo.clone_from(repo_url, repo_path)
-    fetch_pull_repo(repo_path, ref_name)
+    try:
```
remove exception handling here - doesn't make sense

dmytrotkk commented on node_cli/utils/git_utils.py:
```diff
@@ -52,11 +52,21 @@ def sync_repo(repo_url: str, repo_path: str, ref_name: str) -> None:
-        repo.remotes.origin.pull()
+    """Fetch latest changes and checkout/pull specific git reference."""
+
+    try:
```
`try/catch` can be dropped

Masterix0 committed:
Addresses PR #839 comments.

Masterix0 committed:
Removes unused helper import.
